aggregator/election_mgr.go

```diff
@@ -0,0 +1,544 @@
+	mgr.RUnlock()
+	return electionState
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	_, watch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return fmt.Errorf("error creating watch when resigning: %v", err)
+	}
+	defer watch.Close()
+
+	resignAction := resignAction{errCh: make(chan error, 1)}
+	select {
+	case mgr.resignCh <- resignAction:
+	default:
+		return errResignActionAlreadyEnqueued
+	}
+
+	for {
+		select {
+		case <-watch.C():
+			if goalState := watch.Get().(goalState); goalState.state == FollowerState {
+				return nil
+			}
+		case <-ctx.Done():
+			mgr.metrics.resignTimeout.Inc(1)
+			mgr.logError("resign error", ctx.Err())
+			return ctx.Err()
+		case err := <-resignAction.errCh:
+			mgr.metrics.resignErrors.Inc(1)
+			mgr.logError("resign error", err)
+			return err
+		}
+	}
+}
+
+func (mgr *electionManager) Close() error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerOpen {
+		return errElectionManagerExpectOpen
+	}
+	close(mgr.doneCh)
+	mgr.state = electionManagerClosed
+	return nil
+}
+
+func (mgr *electionManager) watchStateChanges(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+			mgr.processGoalState(watch.Get().(goalState))
+		case action := <-mgr.resignCh:
+			mgr.resign(action)
+		}
+	}
+}
+
+func (mgr *electionManager) processGoalState(goalState goalState) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	currState := mgr.electionState
+	newState := goalState.state
+	if currState == newState {
+		return
+	}
+	if currState == FollowerState && newState == PendingFollowerState {
+		return
+	}
+	mgr.electionState = newState
+	mgr.metrics.stateChanges.Inc(1)
+	mgr.logger.Info(fmt.Sprintf("election state changed from %v to %v", currState, newState))
+}
+
+func (mgr *electionManager) verifyPendingFollower(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+		}
+
+		currState := mgr.goalStateWatchable.Get().(goalState)
+		if currState.state != PendingFollowerState {
+			continue
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), mgr.changeVerifyTimeout)
+		continueFn := func(int) bool {
+			select {
+			case <-ctx.Done():
+				return false
+			default:
+				return true
+			}
+		}
+
+		var (
+			knownLeader bool
+			isLeader    bool
+		)
+		mgr.changeRetrier.AttemptWhile(continueFn, func() error {
+			leader, err := mgr.leaderService.Leader(mgr.electionKey)
+			if err != nil {
+				knownLeader = false
+				mgr.metrics.verifyLeaderErrors.Inc(1)
+				mgr.logError("error determining the leader", err)
+				return err
+			}
+
+			knownLeader = true
+			if leader == mgr.leaderValue {
+				isLeader = true
+				mgr.metrics.verifyLeaderNotChanged.Inc(1)
+				mgr.logError("leader has not changed", errLeaderNotChanged)
+				return errLeaderNotChanged
+			}
+			isLeader = false
+			return nil
+		})
+		cancel()
```

Comment:
Apparently cancel is used to release resources associated with contexts in addition to cancelling them. If you don't call `cancel` explicitly, intellisense complains.